hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceManager.java

```diff
@@ -641,8 +641,8 @@ public void removePeer(String id) {
             "and that replication on this peer may not be caught up. peerId=" + id);
       }
       for (ReplicationSourceInterface toRemove : srcToRemove) {
-        toRemove.terminate(terminateMessage);
         closeQueue(toRemove);
+        toRemove.terminate(terminateMessage);
```

Comment:
I'm thinking of ways to do a proper cleanup. My only concern here is this may cause some other problem in the future if we rely on a specific order, let me get back on this.

Reply:
@bharathv The reordering doesn't impact anything IMO since if any of the two calls fail `closeQueue(toRemove);` or `toRemove.terminate(terminateMessage);`, it's the same thing because peer removed call is never retried. 
However, please feel free to suggest anything better.

Reply:
Order matters here... We have fixed the order problems a lot in the past, so give me some time to review the code before merging.

Thanks.

Reply:
Spoke with @sandeepvinayak. So the issue here was introduced by HBASE-25583. That patch introduced a peerRemoved() cleanup call from inside of ReplicationSourceShipperThread when any ZK call to update log position throws NoNodeException, meaning the peer was removed by some other process.

(Ideally the peer should eventually get cleaned up due to a notification from ZK but we have repeatedly observed that there are occasional missed ZK notifications resulting in a weird ZK state, so the patch aggressively cleans up the peer state the first time it sees this exception).

Now the problem is that peerRemoved() attempts to interrupt and cleanup the ReplicationSourceShipperThread from which it is being called and that is aborting the thread half way through and not performing the cleanup. Ideally that shouldn't happen but JVM has some weird semantics about thread interrupting itself, sigh.

So the right way of fixing it is to keep the existing cleanup order as is (which is correct) but instead _not_ call peer cleanup routine from inside shipper/other replication threads. Currently the way the code is structured, there is no easy way (I can think of) to schedule a peer remove event without significant plumbing. A brute force way is to schedule it in the top level executor in ReplicationSourceManager, is there a better way?

Reply:
calling `peerRemoved` was the best thing we thought of calling once we confirm the peer has been removed while handling `NoNodeException`. However, it's not doing the expected thing and doing half things. One alternative I can think of is having a new function other peer Removed which will be called just for this handling and doing the right cleanup and we can terminate the thread at the end of the function.

Reply:
So this is for branch-1 only? I can not recall the exact logic for branch-1 as it has been changed a lot on branch-2+, as now we do not depend on zk notification...

So just go ahead if you guys find the root cause.

Thanks.

Reply:
>  One alternative I can think of is having a new function other peer Removed which will be called just for this handling and doing the right cleanup and we can terminate the thread at the end of the function.

You have a patch for this? Only concern is that creates two versions of "peerRemoved()" functions and they can diverge, if we can avoid that some way, we are good to go.

Reply:
@bharathv  can you review it again? I modified the patch to call peerRemoved from a separate thread instead if it comes from a replication source.